buildworker.rst

```diff
@@ -111,7 +111,64 @@ machine reboots:
 
 For Linux:
 
-    * Add the following line to ``/etc/crontab``::
+    * For systemd based distributions, you can create a systemd unit file in order
+      to manage the service. Create the unit file named ``buildbot-worker.service``
+      under ``/etc/systemd/system`` and change the paths according to where the
+      buildbot-worker binary resides. You can verify its location by running
+      ``which buildbot-worker``. If you installed the buildbot-worker through
+      your package manager it would be::
+
+         [Unit]
+         Description=Buildbot Worker
+         Wants=network.target
+         After=network.target
+
+         [Service]
+         Type=forking
+         PIDFile=/home/buildbot/buildarea/twistd.pid
+         WorkingDirectory=/home/buildbot/buildarea
+         ExecStart=/usr/bin/buildbot-worker start
+         ExecReload=/usr/bin/buildbot-worker restart
+         ExecStop=/usr/bin/buildbot-worker stop
+         Restart=always
+         User=buildbot
+
+         [Install]
+         WantedBy=multi-user.target
+
+      If you installed the buildbot-worker through pip, the systemd unit
+      file will look like this::
+
+        [Unit]
+        Description=Buildbot Worker
+        Wants=network.target
+        After=network.target
+
+        [Service]
+        Type=forking
+        PIDFile=/home/buildbot/buildarea/twistd.pid
+        WorkingDirectory=/home/buildbot/buildarea
+        ExecStart=/usr/local/bin/buildbot-worker start
+        ExecReload=/usr/local/bin/buildbot-worker restart
+        ExecStop=/usr/local/bin/buildbot-worker stop
+        Restart=always
+        User=buildbot
+
+        [Install]
+        WantedBy=multi-user.target
+
+      Then run ``systemctl enable --now buildbot-worker.service`` to start the service
+      and have it activated on every reboot.
+
+      Note that using a systemd unit file, might produce some selinux warnings on systems
+      where the enforcing mode is enabled, usually related to the twistd.pid file.
+      If the service fails to start, you should check the output of
+      ``systemctl status buildbot-worker.service`` as well as the
+      ``/var/log/audit/audit.log`` file for potential issues and remedies.
```

Comment:
``ausearch -m AVC``

Reply:
I usually use `sealert -a /var/log/audit/audit.log` which provides messages with the warning and how to fix them, but I thought it could be too specific for the devguide.

Reply:
Is this really going to be an issue? Unless you define a policy the buildbot daemon should run in ``unconfined_t`` and not be affected by SELinux policy restrictions.

Reply:
It was always affected on all the buildbots I've set up (Fedora and RHEL). Basically selinux prevents systemd's open and read access for the twistd.pid file.

Reply:
So after following the instructions on the buildbot worker page to set it up and creating the systemd unit files, when trying to start the service, initially this error message is show (date/time is trimmed out):

`
Following twistd.log until startup finished..
AVC avc:  denied  { read } for  pid=1 comm="systemd" name="twistd.pid" dev="dm-0" ino=25387713 scontext=system_u:system_r:init_t:s0 tcontext=system_u:object_r:user_home_t:s0 tclass=file permissive=0

AVC avc:  denied  { read } for  pid=1 comm="systemd" name="twistd.pid" dev="dm-0" ino=25387713 scontext=system_u:system_r:init_t:s0 tcontext=system_u:object_r:user_home_t:s0 tclass=file permissive=0

AVC avc:  denied  { read } for  pid=1 comm="systemd" name="twistd.pid" dev="dm-0" ino=25387713 scontext=system_u:system_r:init_t:s0 tcontext=system_u:object_r:user_home_t:s0 tclass=file permissive=0

localhost systemd[1]: buildbot-worker.service: Can't convert PID files /home/buildbot/buildarea/twistd.pid O_PATH file descriptor to proper file descriptor: Permission denied

localhost systemd[1]: buildbot-worker.service: Can't convert PID files /home/buildbot/buildarea/twistd.pid O_PATH file descriptor to proper file descriptor: Permission denied

localhost setroubleshoot[1489084]: SELinux is preventing systemd from read access on the file twistd.pid. For complete SELinux messages run: sealert -l 57d3c874-be75-46f6-82c9-3680ac54002c

localhost python3[1489084]: SELinux is preventing systemd from read access on the file twistd.pid.#012#012*****  Plugin catchall (100. confidence) suggests   **************************#012#012
If you believe that systemd should be allowed read access on the twistd.pid file by default.#012Then you should report this as a bug.#012You can generate a local policy module to allow this access.#012Do#012allow this access for now by executing:#012# ausearch -c 'systemd' --raw | audit2allow -M my-systemd#012# semodule -X 300 -i my-systemd.pp#012

localhost setroubleshoot[1489084]: SELinux is preventing systemd from read access on the file twistd.pid. For complete SELinux messages run: sealert -l 57d3c874-be75-46f6-82c9-3680ac54002c

localhost python3[1489084]: SELinux is preventing systemd from read access on the file twistd.pid.#012#012*****  Plugin catchall (100. confidence) suggests   **************************#012#012If you believe that systemd should be allowed read access on the twistd.pid file by default.#012Then you should report this as a bug.#012You can generate a local policy module to allow this access.#012Do#012allow this access for now by executing:#012# ausearch -c 'systemd' --raw | audit2allow -M my-systemd#012# semodule -X 300 -i my-systemd.pp#012
`
After adjusting the policy, another error comes up:

`
Following twistd.log until startup finished..

localhost audit[1]: AVC avc:  denied  { open } for  pid=1 comm="systemd" path="/home/buildbot/buildarea/twistd.pid" dev="dm-0" ino=25387713 scontext=system_u:system_r:init_t:s0 tcontext=system_u:object_r:user_home_t:s0 tclass=file permissive=0

localhost systemd[1]: buildbot-worker.service: Can't convert PID files /home/buildbot/buildarea/twistd.pid O_PATH file descriptor to proper file descriptor: Permission denied

localhost systemd[1]: buildbot-worker.service: Can't convert PID files /home/buildbot/buildarea/twistd.pid O_PATH file descriptor to proper file descriptor: Permission denied

localhost audit[1]: AVC avc:  denied  { open } for  pid=1 comm="systemd" path="/home/buildbot/buildarea/twistd.pid" dev="dm-0" ino=25387713 scontext=system_u:system_r:init_t:s0 tcontext=system_u:object_r:user_home_t:s0 tclass=file permissive=0

localhost setroubleshoot[8714]: SELinux is preventing systemd from open access on the file /home/buildbot/buildarea/twistd.pid. For complete SELinux messages run: sealert -l 3d54c639-fea4-4a18-be47-c5fe1d57a02a

localhost python3[8714]: SELinux is preventing systemd from open access on the file /home/buildbot/buildarea/twistd.pid.#012#012*****  Plugin catchall (100. confidence) suggests   **************************#012#012If you believe that systemd should be allowed open access on the twistd.pid file by default.#012Then you should report this as a bug.#012You can generate a local policy module to allow this access.#012Do#012allow this access for now by executing:#012# ausearch -c 'systemd' --raw | audit2allow -M my-systemd#012# semodule -X 300 -i my-systemd.pp#012`

After allowing that as well, the service starts normally.